Feature/document size check

StephenMakrogianni opened the pull request:
Elastic has a maximum document size they will index. Added a config variable to prevent trying to index documents that will be rejected by search providers. A warning is thrown with the details of any document that is bigger than the defined size.

StephenMakrogianni committed:
Add document max size variable

Some search vendors have limits on the maximum document size that can be indexed. Enter $document_max_size, this means we catch documents that exceed this limit before we send them to the search vendor.

StephenMakrogianni committed:
Add documentation for document_max_size config variable

unclecheese commented on src/Service/AppSearch/AppSearchService.php:
```diff
@@ -82,6 +82,19 @@ public function addDocuments(array $items): BatchDocumentInterface
             $fields = $this->getBuilder()->toArray($item);
+
+            $documentMaxSize = $this->getConfiguration()->getDocumentMaxSize();
+            if ($documentMaxSize  && strlen(serialize($fields)) >= $documentMaxSize) {
```
Should we use `json_encode($fields)` here to make it more analogous to how Elastic will store it?

StephenMakrogianni replied:
Yeah good point. I'll change it.

unclecheese reviewed (changes requested):
Looks good. One of the things I'm trying to avoid is each indexing implementation sharing boilerplate code. When that happens, we should try to encapsulate it into `DocumentBuilder`, which is the service that says, "no matter what search service you're using, these things need to happen". (Adding ID fields, for instance).

It seems probable that all indexing services will have a max document size, so I think it _does_ belong in `IndexConfiguration` (generic, cross-service settings), however, what that limit is now becomes a concern of the indexing service.

Let's expand the `IndexingInterface` to require a `getMaxDocumentSize(): int` method. `AppSearchService` will add that method. Fallback on a native config value.

Then, in `DocumentBuilder`, the `toArray()` will do the validation. It can then trust that `IndexingInterface` will tell it what the document size needs to be.

Then, any further implementations -- Algolia, etc., can just provide `getMaxDocumentSize()` and they don't have to worry about writing this same validation because it's all handled in `DocumentBuilder`.

With regard to throwing a warning, I think that's appropriate, but I just want to make sure we're not hiding these bypassed records from the end user. If documents aren't indexing, that's important information. It could be a critical document update that it supposed to remove sensitive information, for instance.

That might be a bigger conversation, though -- how we handle failures like that.

Make sense?

unclecheese commented on src/Service/AppSearch/AppSearchService.php:
```diff
@@ -82,6 +82,19 @@ public function addDocuments(array $items): BatchDocumentInterface
+                    $fields['source_class'],
+                    $fields['record_id'],
+                    $documentMaxSize
+                ), E_USER_WARNING);
```
Let's upgrade this to an `IndexServiceException`, because we'll need to throw the job into a failure state to get retries.

StephenMakrogianni replied:
Sweet i'll swap it out mate

StephenMakrogianni replied:
Using `IndexServiceException` causes the bulk reindex to break. If works fine for a single record but when bulk reindexing the exception stops the reindex.

StephenMakrogianni commented:
Yup that makes sense mate. I'll make the adjustments now.

@unclecheese  In regards to the error handling, agreed we may need to discuss how we want to handle that. I agree that the user needs to be made aware if a particular record doesn't get indexed. How we do that i'm not sure. Maybe create a index report tab in the sites settings or a custom model admin. This would mean we could store more detail about what has been indexed, when and any errors that occurred. It may be over kill, just an idea at this stage.

StephenMakrogianni committed:
Address PR feedback

Made max document size more robust by validating the config variable in DocumentBuilder

StephenMakrogianni committed:
Tidy up docblocks and updated function to correctly get max document size

StephenMakrogianni committed:
Add Todo in the max document size validation

Commit:
Revisions to isolate to DocumentBuilder

StephenMakrogianni commented on _config/config.yml:
```diff
@@ -46,3 +46,4 @@ SilverStripe\Core\Injector\Injector:
     constructor:
       configuration: '%$SilverStripe\SearchService\Service\IndexConfiguration'
       registry: '%$SilverStripe\SearchService\Service\DocumentFetchCreatorRegistry'
+      service: '%$SilverStripe\SearchService\Interfaces\IndexingInterface'
```
This appears to cause an infinite loop when doing dev/build?flush. I think it's related to: https://github.com/silverstripe/silverstripe-search-service/blob/869006d62fd7423ddc4bd25f4add1f5ea7054d7d/_config/appsearch.yml#L14
Any ideas how to resolve this mate?

StephenMakrogianni committed:
Removed unneeded interface. Truncate long fields in so they can be indexed.

- Removed IndexingInterface from the constructor.
- Truncate long fields that are going to be indexed. This means that we can still index some of the record.

StephenMakrogianni committed:
Change truncation function so the array index is preserved. This is
because the schema will display a warning icon which I believe means
it's borked

StephenMakrogianni committed:
Fix coding standards issue

chillu committed:
DOCS Fixed var name

andrewandante commented:
Hi team - anything else required to get this merged into the main branch?

StephenMakrogianni commented:
@andrewandante not as far as i'm aware. @unclecheese what are your thoughts?

unclecheese commented:
Looks good. Thanks, everyone.
